// @mongosh/shell-api/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADMIN_DB = exports.namespaceInfo = exports.asPrintable = exports.shellApiType = exports.CURSOR_FLAGS = exports.ALL_API_VERSIONS = exports.ALL_PLATFORMS = exports.ALL_TOPOLOGIES = exports.ALL_SERVER_VERSIONS = exports.Topologies = exports.ServerVersions = void 0;
var ServerVersions;
(function (ServerVersions) {
    ServerVersions["latest"] = "999.999.999";
    ServerVersions["earliest"] = "0.0.0";
})(ServerVersions = exports.ServerVersions || (exports.ServerVersions = {}));
var Topologies;
(function (Topologies) {
    Topologies["ReplSet"] = "ReplSet";
    Topologies["Standalone"] = "Standalone";
    Topologies["Sharded"] = "Sharded";
    Topologies["LoadBalanced"] = "LoadBalanced";
})(Topologies = exports.Topologies || (exports.Topologies = {}));
exports.ALL_SERVER_VERSIONS = [ServerVersions.earliest, ServerVersions.latest];
exports.ALL_TOPOLOGIES = [Topologies.ReplSet, Topologies.Sharded, Topologies.LoadBalanced, Topologies.Standalone];
exports.ALL_PLATFORMS = ['Compass', 'Browser', 'CLI'];
exports.ALL_API_VERSIONS = [0, Infinity];
exports.CURSOR_FLAGS = {
    2: 'tailable',
    4: 'SlaveOk',
    8: 'oplogReplay',
    16: 'noTimeout',
    32: 'awaitData',
    64: 'exhaust',
    128: 'partial'
};
exports.shellApiType = Symbol.for('@@mongosh.shellApiType');
exports.asPrintable = Symbol.for('@@mongosh.asPrintable');
exports.namespaceInfo = Symbol.for('@@mongosh.namespaceInfo');
exports.ADMIN_DB = 'admin';
