// @mongosh/shell-api/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getShellApiType = exports.toShellResult = exports.Topologies = exports.ServerVersions = exports.ShellApi = exports.signatures = exports.UpdateResult = exports.ReplicaSet = exports.Shard = exports.Mongo = exports.InsertOneResult = exports.InsertManyResult = exports.DeleteResult = exports.CommandResult = exports.BulkWriteResult = exports.ShellInstanceState = exports.Help = exports.ExplainableCursor = exports.Explainable = exports.Collection = exports.Database = exports.CursorIterationResult = exports.Cursor = exports.RunCommandCursor = exports.AggregationCursor = void 0;
const aggregation_cursor_1 = __importDefault(require("./aggregation-cursor"));
exports.AggregationCursor = aggregation_cursor_1.default;
const run_command_cursor_1 = __importDefault(require("./run-command-cursor"));
exports.RunCommandCursor = run_command_cursor_1.default;
const collection_1 = __importDefault(require("./collection"));
exports.Collection = collection_1.default;
const cursor_1 = __importDefault(require("./cursor"));
exports.Cursor = cursor_1.default;
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
const explainable_1 = __importDefault(require("./explainable"));
exports.Explainable = explainable_1.default;
const explainable_cursor_1 = __importDefault(require("./explainable-cursor"));
exports.ExplainableCursor = explainable_cursor_1.default;
const help_1 = __importDefault(require("./help"));
exports.Help = help_1.default;
const shell_instance_state_1 = __importDefault(require("./shell-instance-state"));
exports.ShellInstanceState = shell_instance_state_1.default;
const shard_1 = __importDefault(require("./shard"));
exports.Shard = shard_1.default;
const replica_set_1 = __importDefault(require("./replica-set"));
exports.ReplicaSet = replica_set_1.default;
const shell_api_1 = __importDefault(require("./shell-api"));
exports.ShellApi = shell_api_1.default;
const result_1 = require("./result");
Object.defineProperty(exports, "BulkWriteResult", { enumerable: true, get: function () { return result_1.BulkWriteResult; } });
Object.defineProperty(exports, "CommandResult", { enumerable: true, get: function () { return result_1.CommandResult; } });
Object.defineProperty(exports, "CursorIterationResult", { enumerable: true, get: function () { return result_1.CursorIterationResult; } });
Object.defineProperty(exports, "DeleteResult", { enumerable: true, get: function () { return result_1.DeleteResult; } });
Object.defineProperty(exports, "InsertManyResult", { enumerable: true, get: function () { return result_1.InsertManyResult; } });
Object.defineProperty(exports, "InsertOneResult", { enumerable: true, get: function () { return result_1.InsertOneResult; } });
Object.defineProperty(exports, "UpdateResult", { enumerable: true, get: function () { return result_1.UpdateResult; } });
const mongo_1 = __importDefault(require("./mongo"));
exports.Mongo = mongo_1.default;
const decorators_1 = require("./decorators");
Object.defineProperty(exports, "signatures", { enumerable: true, get: function () { return decorators_1.signatures; } });
Object.defineProperty(exports, "toShellResult", { enumerable: true, get: function () { return decorators_1.toShellResult; } });
Object.defineProperty(exports, "getShellApiType", { enumerable: true, get: function () { return decorators_1.getShellApiType; } });
const enums_1 = require("./enums");
Object.defineProperty(exports, "Topologies", { enumerable: true, get: function () { return enums_1.Topologies; } });
Object.defineProperty(exports, "ServerVersions", { enumerable: true, get: function () { return enums_1.ServerVersions; } });
