// @mongosh/logging/lib/analytics-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToggleableAnalytics = exports.NoopAnalytics = void 0;
class NoopAnalytics {
    identify(_info) { }
    track(_info) { }
}
exports.NoopAnalytics = NoopAnalytics;
class ToggleableAnalytics {
    constructor(target = new NoopAnalytics()) {
        this._queue = [];
        this._state = 'paused';
        this._target = target;
    }
    identify(...args) {
        this._validateArgs(args);
        switch (this._state) {
            case 'enabled':
                this._target.identify(...args);
                break;
            case 'paused':
                this._queue.push(['identify', args]);
                break;
            default:
                break;
        }
    }
    track(...args) {
        this._validateArgs(args);
        switch (this._state) {
            case 'enabled':
                this._target.track(...args);
                break;
            case 'paused':
                this._queue.push(['track', args]);
                break;
            default:
                break;
        }
    }
    enable() {
        if (this._pendingError) {
            throw this._pendingError;
        }
        this._state = 'enabled';
        const queue = this._queue;
        this._queue = [];
        for (const entry of queue) {
            if (entry[0] === 'identify')
                this.identify(...entry[1]);
            if (entry[0] === 'track')
                this.track(...entry[1]);
        }
    }
    disable() {
        this._state = 'disabled';
        this._pendingError = undefined;
        this._queue = [];
    }
    pause() {
        this._state = 'paused';
    }
    _validateArgs([firstArg]) {
        var _a;
        if (!('userId' in firstArg && firstArg.userId) &&
            !('anonymousId' in firstArg && firstArg.anonymousId)) {
            const err = new Error('Telemetry setup is missing userId or anonymousId');
            switch (this._state) {
                case 'enabled':
                    throw err;
                case 'paused':
                    (_a = this._pendingError) !== null && _a !== void 0 ? _a : (this._pendingError = err);
                    break;
                default:
                    break;
            }
        }
    }
}
exports.ToggleableAnalytics = ToggleableAnalytics;
